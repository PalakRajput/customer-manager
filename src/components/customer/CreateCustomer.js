import React, { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { createCustomerAction, getCustomerAction, updateCustomer } from '../../redux/actions/CustomerActions';
import { useLocation } from 'react-router-dom';

const CreateCustomer = () => {
    const [firstName, setFirstName] = useState("");
    const [lastName, setLastName] = useState("");
    const [city, setCity] = useState("");
    const [error, setError] = useState("");
    const dispatch = useDispatch();
    const location = useLocation();
    const successMessage = useSelector(state => state.customer.successMessage)
     
    useEffect(() => {
        //for updating the customer
        if(location.state !== null){
            const {customer} = location.state
            setFirstName(customer.fname)
            setLastName(customer.lname)
            setCity(customer.city)
        }else{
            setFirstName("")
            setLastName("")
            setCity("")
        }
    },[location.state])
    const handleClick = (e) => {
        if ((firstName !== "" && firstName !== null) && (lastName !== "" && lastName !== null) && (city !== "" && city !== null)) {
            setError("")
            if(location.state !== null){
                //action to update customer
                dispatch(updateCustomer({fname: firstName, lname: lastName, city, id: location.state.id}, updateCustomerSuccess))
            }else{
                dispatch(createCustomerAction({ fname: firstName, lname: lastName, city }, updateCustomerSuccess))
            }
            setFirstName("");
            setLastName("");
            setCity("");
        } else {
            setError("Please enter data in all fields.")
        }
    }
    const updateCustomerSuccess = () => {
        dispatch(getCustomerAction())
    }
    return (
        <div className='container' >
            <div className='row' style={{ marginTop: '50px' }}>
                <div className='col-2'>
                    <label htmlFor='fname' style={{ fontWeight: 'bold' }}> First Name: </label>
                    <input data-testid="fname" type="text" id="fname" value={firstName} onChange={(e) => setFirstName(e.target.value)} placeholder="First Name" style={{ height: '40px', borderRadius: '5px' }}></input>
                </div>
                <div className='col-1'></div>
                <div className='col-2'>
                    <label htmlFor='lname' style={{ fontWeight: 'bold' }}> Last Name: </label>
                    <input data-testid="lname" type="text" id="lname" value={lastName} onChange={(e) => setLastName(e.target.value)} placeholder="Last Name" style={{ height: '40px', borderRadius: '5px' }}></input>
                </div>
                <div className='col-1'></div>
                <div className='col-2'>
                    <label htmlFor='city' style={{ fontWeight: 'bold' }}> City: </label>
                    <input data-testid="city" type="text" id="city" value={city} onChange={(e) => setCity(e.target.value)} placeholder="City" style={{ height: '40px', borderRadius: '5px' }}></input>
                </div>
                <div className='col-1'></div>
                <div className='col-2'>
                    <br />
                    <button data-testid="addButton" disabled={!firstName || !lastName || !city} className='btn btn-primary' style={{ width: 'max-content', alignItems: 'center', height: '40px' }} onClick={handleClick}>{location.state !== null ? 'Update Customer' : 'Add Customer'}</button>
                </div>
            </div>

            <br />
            {error ? <span className='text-danger font-weight-bold'>{error}</span> : null}
            {/* react-toastify to hide the message after 2secs */}
            {successMessage ? <span className='text-success font-weight-bold'>{successMessage}</span> : null}
        </div>
    )

}

export default CreateCustomer;