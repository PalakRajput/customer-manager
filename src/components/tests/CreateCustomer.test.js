import { render, cleanup, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import thunk from "redux-thunk";
import CreateCustomer from '../customer/CreateCustomer';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store'
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)
const store = mockStore({
    customer:{
        customers:[],
        successMessage: "",
        error:""
    }
})

afterEach(cleanup)

describe("Create Customer", () => {
    it (" should render correctly ", () => {
           const {asFragment} =  render(<MemoryRouter><Provider store={store}><CreateCustomer /></Provider></MemoryRouter>)
           expect(asFragment(<CreateCustomer />)).toMatchSnapshot();
    })
})

it (" should have initial first name as empty string ", () => {
    const {getByTestId} =  render(<MemoryRouter><Provider store={store}><CreateCustomer /></Provider></MemoryRouter>)
    expect(getByTestId('fname')).toHaveDisplayValue("");
})
it (" should have initial last name as empty string ", () => {
    const {getByTestId} =  render(<MemoryRouter><Provider store={store}><CreateCustomer /></Provider></MemoryRouter>)
    expect(getByTestId('fname')).toHaveDisplayValue("");
})
it (" should have initial city as empty string ", () => {
    const {getByTestId} =  render(<MemoryRouter><Provider store={store}><CreateCustomer /></Provider></MemoryRouter>)
    expect(getByTestId('fname')).toHaveTextContent("");
})
it (" should not have button disabled initially ", () => {
    const {getByTestId} =  render(<MemoryRouter><Provider store={store}><CreateCustomer /></Provider></MemoryRouter>)
    expect(getByTestId('addButton')).toBeDisabled();
})

// it('tests the input text value after onChange event ', () => {
//     const {getByPlaceholderText} = render(<MemoryRouter><Provider store={store}><CreateCustomer /></Provider></MemoryRouter>); 
//     // fireEvent.change(getByTestId('fname'), {target:{value:'new val'}})
//     const input = getByPlaceholderText('First Name');
//     userEvent.type(input, "new val")
//     expect(input).toHaveValue("new val")
//   });