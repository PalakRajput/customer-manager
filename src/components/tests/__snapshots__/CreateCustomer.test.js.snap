// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Customer  should render correctly  1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
      style="margin-top: 50px;"
    >
      <div
        class="col-2"
      >
        <label
          for="fname"
          style="font-weight: bold;"
        >
           First Name: 
        </label>
        <input
          data-testid="fname"
          id="fname"
          placeholder="First Name"
          style="height: 40px; border-radius: 5px;"
          type="text"
          value=""
        />
      </div>
      <div
        class="col-1"
      />
      <div
        class="col-2"
      >
        <label
          for="lname"
          style="font-weight: bold;"
        >
           Last Name: 
        </label>
        <input
          data-testid="lname"
          id="lname"
          placeholder="Last Name"
          style="height: 40px; border-radius: 5px;"
          type="text"
          value=""
        />
      </div>
      <div
        class="col-1"
      />
      <div
        class="col-2"
      >
        <label
          for="city"
          style="font-weight: bold;"
        >
           City: 
        </label>
        <input
          data-testid="city"
          id="city"
          placeholder="City"
          style="height: 40px; border-radius: 5px;"
          type="text"
          value=""
        />
      </div>
      <div
        class="col-1"
      />
      <div
        class="col-2"
      >
        <br />
        <button
          class="btn btn-primary"
          data-testid="addButton"
          disabled=""
          style="align-items: center; height: 40px;"
        >
          Add Customer
        </button>
      </div>
    </div>
    <br />
  </div>
</DocumentFragment>
`;
